Jetchat/app/src/main/java/com/example/compose/jetchat/conversation/RecordButton.kt

```diff
@@ -0,0 +1,165 @@
+import androidx.compose.ui.platform.LocalDensity
+import androidx.compose.ui.unit.dp
+import kotlin.math.abs
+import kotlinx.coroutines.launch
+
+val swipeToCancelThreshold: () -> Float
+    @Composable get() = with(LocalDensity.current) { { 200.dp.toPx() } }
+
+private val verticalThreshold: () -> Float
+    @Composable get() = with(LocalDensity.current) { { 80.dp.toPx() } }
+
+@OptIn(ExperimentalMaterial3Api::class)
+@Composable
+fun RecordButton(
+    recording: Boolean,
+    swipeOffset: MutableState<Float>,
+    onStartRecording: () -> Boolean,
+    onFinishRecording: () -> Unit,
+    onCancelRecording: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    val transition = updateTransition(targetState = recording, label = "record")
+    val scale = transition.animateFloat(
+        transitionSpec = { spring(Spring.DampingRatioMediumBouncy, Spring.StiffnessLow) },
+        label = "record-scale",
+        targetValueByState = { rec -> if (rec) 2f else 1f }
+    )
+    val containerAlpha = transition.animateFloat(
+        transitionSpec = { tween(2000) },
+        label = "record-scale",
+        targetValueByState = { rec -> if (rec) 1f else 0f }
+    )
+    val iconColor = transition.animateColor(
+        transitionSpec = { tween(200) },
+        label = "record-scale",
+        targetValueByState = { rec ->
+            if (rec) contentColorFor(LocalContentColor.current)
+            else LocalContentColor.current
+        }
+    )
+
+    Box {
+        // Background during recording
+        Box(
+            Modifier
+                .matchParentSize()
+                .aspectRatio(1f)
+                .graphicsLayer {
+                    alpha = containerAlpha.value
+                    scaleX = scale.value; scaleY = scale.value
+                }
+                .clip(CircleShape)
+                .background(LocalContentColor.current)
+        )
+        val scope = rememberCoroutineScope()
+        val tooltipState = remember { RichTooltipState() }
+        RichTooltipBox(
+            text = { Text("Touch and hold to record") },
+            tooltipState = tooltipState
+        ) {
+            Icon(
+                Icons.Default.Mic,
+                contentDescription = "Record voice message",
+                tint = iconColor.value,
+                modifier = modifier
+                    .sizeIn(minWidth = 56.dp, minHeight = 6.dp)
+                    .padding(18.dp)
+                    .recordGesture(
+                        swipeOffset,
+                        swipeToCancelThreshold(),
+                        verticalThreshold(),
+                        { scope.launch { tooltipState.show() } },
+                        onStartRecording,
+                        onFinishRecording,
+                        onCancelRecording,
+                    )
+            )
+        }
+    }
+}
+
+private fun Modifier.recordGesture(
+    horizontalSwipeProgress: MutableState<Float>,
+    swipeToCancelThreshold: Float,
```

Comment:
could you pass those functions as defaults? Feels like this is exposing quite a bit of extra detail that may not be needed. Or is this something that could be taken from the container size? How does it work with a larger screen?

Reply:
The values should be fixed no matter the size of the screen (it's basically just large enough values to make sure you're being intentional with your cancel affordance).

So what I was struggling with here is that you need screen density to get those fixed values. Now I can make my modifier stateful using `composed` but I don't think we should be doing that.